[3기 mmmindy] 자바스크립트 계산기 미션

mmindy opened the pull request:
## 🎯 기능 요구사항

- [X] 2개의 숫자에 대해 덧셈이 가능하다.
- [X] 2개의 숫자에 대해 뺄셈이 가능하다.
- [X] 2개의 숫자에 대해 곱셈이 가능하다.
- [X] 2개의 숫자에 대해 나눗셈이 가능하다.
- [X] AC(All Clear)버튼을 누르면 0으로 초기화 한다.
- [X] 숫자는 한번에 최대 3자리 수까지 입력 가능하다.
- [X] 계산 결과를 표현할 때 소수점 이하는 버림한다.


## 📊 테스트 요구사항

- [X] 기능 요구사항에 제시된 7개의 항목에 대해 테스트 케이스를 만든다.

exemfe3 committed:
[chore] 계산기 기능 구현
##########
- [X] 2개의 숫자에 대해 덧셈이 가능하다.
- [X] 2개의 숫자에 대해 뺄셈이 가능하다.
- [X] 2개의 숫자에 대해 곱셈이 가능하다.
- [X] 2개의 숫자에 대해 나눗셈이 가능하다.
- [X] AC(All Clear)버튼을 누르면 0으로 초기화 한다.
- [X] 숫자는 한번에 최대 3자리 수까지 입력 가능하다.
- [X] 계산 결과를 표현할 때 소수점 이하는 버림한다.

exemfe3 committed:
[test/refactor] 테스트 코드 & 리팩토링
##########
테스트 코드 & 리팩토링

lsw1164 commented on src/js/index.js:
```diff
@@ -0,0 +1,5 @@
+import Calculator from './calculator.js';
+
+const container = document.body.getElementsByClassName('calculator')[0];
```
``document.querySelector('.calculator');`` 로 하시면 더 간략하게 작성할 수 있습니다 :)

mmindy replied:
넵ㅎㅎ! 감사합니다

lsw1164 commented on src/js/common/define.js:
```diff
@@ -0,0 +1,10 @@
+export const OPERATOR = {
+  DIVIDED: '/',
+  MULTIPLIED: 'X',
+  PLUS: '+',
+  MINUS: '-',
+  EQUAL: '=',
+};
+export const MODIFIER = {
+  AC: 'AC',
+};
```
신경 쓰지 못할 때가 많은데, Const를 따로 관리하는 건 언제나 옳다고 생각합니다 👍

lsw1164 commented on cypress/integration/calculator.spec.js:
```diff
@@ -0,0 +1,80 @@
+
+  it('계산 결과를 표현할 때 소수점 이하는 버림한다.', () => {
+    testCalculate(10, 3, OPERATOR.DIVIDED);
+    cy.get('#total').should('not.contain', '.');
```
``cy.get('#total').should('not.contain', '.');`` 좋은 아이디어네요 👍

``cy.get('#total').should('have.text', '3');``으로 검증하면 더 명확한 테스트 케이스가 될 거 같습니다

mmindy replied:
단순히 생각했는데, 피드백 감사합니다

lsw1164 commented on cypress/integration/calculator.spec.js:
```diff
@@ -0,0 +1,80 @@
+    for (let ix = 0; ix < str.length; ix++) {
+      cy.get('.digit').contains(str[ix]).click();
+    }
+  };
```
여러 자리의 수를 받아서 한 번에 클릭해주는 함수를 짜셨군요 👏

zel0rd commented on cypress/integration/calculator.spec.js:
```diff
@@ -0,0 +1,80 @@
+import { OPERATOR, MODIFIER } from "../../src/js/common/define.js";
+
+describe('calculator test', () => {
+  beforeEach(() => {
+    cy.visit('http://127.0.0.1:8888/')
+  })
```
잘 몰라서그러는데요~ 혹시 baseUrl를 설정해줘도 beforeEach를 해줘야되나요~??

mmindy replied:
생략해주어도 되는 걸로 알고 있는데 이부분 저도 다시 확인해봐야 겠네요! 감사합니다!

https://docs.cypress.io/guides/references/best-practices#Setting-a-global-baseUrl

iyhong commented on cypress/integration/calculator.spec.js:
```diff
@@ -0,0 +1,80 @@
+    }
+  };
+
+  const testCalculate = (num1, num2, operator) => {
```
함수로 빼서 반복사용을 줄여주셨네요. 너무 좋은것 같습니다.

zel0rd commented on cypress/integration/calculator.spec.js:
```diff
@@ -0,0 +1,80 @@
+        result = num1 * num2;
+        break;
+      case OPERATOR.DIVIDED:
+        result = Math.floor(num1 / num2);
```
혹시 0으로 나누는 예외처리도 여기서 해주면 어떨까 고민해봤습니다!!

mmindy replied:
넵 맞습니다. 말씀해주신 사항 예외처리가 누락됐는데, 추가해보겠습니다ㅎㅎ 감사합니다

lsw1164 commented on src/js/calculator.js:
```diff
@@ -0,0 +1,117 @@
+      if (isNaN(last)) {
+        valueList.push(digit);
+      } else {
+        const tempNum = +(last.toString() + digit.toString());
```
`` const tempNum = +`${last}${digit}`; `` 표현법은 어떨까요?

mmindy replied:
알고 있던 방법이나 코드를 직접 작성할 때 인지를 잘 못하는데, 제안주신 방법 정말 좋은 것같습니다! 감사합니다

zel0rd commented on cypress/integration/calculator.spec.js:
```diff
@@ -0,0 +1,80 @@
+  };
+
+  const clickNumber = (num) => {
+    const str = num.toString();
+    for (let ix = 0; ix < str.length; ix++) {
+      cy.get('.digit').contains(str[ix]).click();
+    }
```
혹시 고차함수를 이용해서 구현하면 어떨까요?!

mmindy replied:
네 `Array.prototype.map`를 사용하면 좀더 깔끔하게 사용할 수 있겠네요! 감사합니다

iyhong commented on src/js/calculator.js:
```diff
@@ -0,0 +1,117 @@
+    this.total = this.container.getElementsByClassName('total')[0];
+    this.addEvent();
+  }
+  addEvent() {
```
이벤트를 넣어줄때 document.querySelector('.digits') 같은 상위에 주게되면 하위 태그에도 모두 적용되더라구요

mmindy replied:
넵 맞습니다! 이벤트 버블링 활용해서 적용해보겠습니다. 감사합니다ㅎ_ㅎ

lsw1164 commented on src/js/calculator.js:
```diff
@@ -0,0 +1,117 @@
+    this.digits = this.container.getElementsByClassName('digit');
+    this.operations = this.container.getElementsByClassName('operation');
+    this.modifiers = this.container.getElementsByClassName('modifier');
+    this.total = this.container.getElementsByClassName('total')[0];
```
dom 객체 하나만 가져온다면 querySelector를 제안드립니다 :)

mmindy replied:
넵 사용하던 것만 사용하다보니 이부분을 놓쳤습니다. 감사합니다ㅎㅎ

lsw1164 commented on src/js/calculator.js:
```diff
@@ -0,0 +1,117 @@
+    this.addEvent();
+  }
+  addEvent() {
+    for (let ix = 0; ix < this.digits.length; ix++) {
+      this.digits[ix].addEventListener('click', this.clickDigit.bind(this));
+    }
+    for (let ix = 0; ix < this.operations.length; ix++) {
+      this.operations[ix].addEventListener('click', this.clickOperation.bind(this));
+    }
+    for (let ix = 0; ix < this.modifiers.length; ix++) {
+      this.modifiers[ix].addEventListener('click', this.clickModifier.bind(this));
+    }
```
각각의 객체에 이벤트를 직접 바인드 하기보다 Container에 직접 바인드하고 이벤트 버블링을 써보는 건 어떨까요?

https://ko.javascript.info/bubbling-and-capturing

mmindy replied:
코드 작성하면서도 찝찝했는데 이벤트 버블링 사용하는 것이 더 깔끔하게 작성할 수 있을 것 같습니다! 감사합니다

zel0rd commented on cypress/integration/calculator.spec.js:
```diff
@@ -0,0 +1,80 @@
+    cy.get('#total').should('have.text', 0);
+  });
+
+  it('숫자는 한번에 최대 3자리 수까지 입력 가능하다.', () => {
+    clickNumber(111);
+    cy.get('.digit').contains('1').click().then(() => {
+      cy.on('window:alert', (str) => {
+        expect(str).should('not.be.empty');
+      });
+    });
+  });
+
```
alert메시지를 확인하고, 입력된 숫자도 3자리까지인지 확인해보면 더욱 좋지 않을까요??

mmindy replied:
네 맞습니다. 테스트 코드가 낯설다보니 좀 엉성하게 짠 듯합니다. 말씀해주신 사항 참고해서 반영해보겠습니다ㅎ_ㅎ

lsw1164 reviewed (commented):
안녕하세요, 코드 잘봤습니다ㅎㅎ 좋은 아이디어 얻어갑니다! :)
몇가지 코멘트 남겼으니 확인부탁드려요!

iyhong reviewed (commented):
안녕하세요
좋은 코드 잘봤습니다. 부족한 실력이라 좋은 리뷰는 못달아 드렸네요ㅠ
테스트케이스가 요구사항에 있는것만 만드셨던데 
그 외에도 필요한 테스트케이스를만들었다면 조금 더 좋았을것 같습니다.
수고많으셨어요!

zel0rd commented on src/js/calculator.js:
```diff
@@ -0,0 +1,117 @@
+      this.modifiers[ix].addEventListener('click', this.clickModifier.bind(this));
+    }
+  }
+  clickDigit(e) {
+    const digit = +e.target.textContent;
+    const valueList = this.valueList;
+    if (isNaN(digit)) {
+      return;
+    }
+
+    if (!valueList.length) {
+      valueList.push(digit);
+    } else {
+      const last = valueList[valueList.length -1];
+      if (isNaN(last)) {
+        valueList.push(digit);
+      } else {
+        const tempNum = +(last.toString() + digit.toString());
+        if (tempNum > 999) {
+          warnMsg('계산기 최대 입력값은 999입니다');
+        } else {
+          valueList[valueList.length -1] = tempNum;
+        }
+      }
+    }
+    this.setTotal();
+  }
```
if (isNaN(digit)) { return; } 하신 것처럼 중첩 if를 간결하게 해보면 어떨까요???

mmindy replied:
흠 이부분 짜면서도 지저분하다고 생각했는데, 좀더 고민해보겠습니다. 감사합니다!

zel0rd reviewed (commented):
코드리뷰가 많이 낯선데... 덕분에 보고 많이 배웠습니다!! 

감사합니다!! 좋은하루 되세요!!

exemfe3 committed:
[refactor] 피드백 반영
##########
- dom 선택자 간소화(querySelector 사용)
- 이벤트 버블링 활용
- 백틱을 활용하여 string 형변형
- 고차함수 활용(Array.prototype.map)
- 0으로 나눗셈 예외 처리
- 중첩 조건문 간소화
- 소수점 테스트 코드 수정
- 세자리 이상 입력 테스트 코드 수정

exemfe3 committed:
[refactor] 공통 피드백 반영
##########
- dom 선택자 추상화
